lib/core/src/Ouroboros/Network/Client/Wallet.hs

```diff
@@ -306,51 +292,67 @@ chainSyncWithBlocks tr fromTip queue responseBuffer =
 
     -- | Keep only blocks from the list that are before or exactly at the given
     -- point.
-    rollback :: Point block -> [block] -> [block]
-    rollback pt = filter (\b -> At (blockSlot b) <= pointSlot pt)
+    rollbackBuffer :: Point block -> [block] -> [block]
+    rollbackBuffer pt = filter (\b -> At (blockSlot b) <= pointSlot pt)
+
+    clientStNegotiateIntersection
+        :: m (P.ClientPipelinedStIdle 'Z block (Point block) (Tip block) m Void)
+    clientStNegotiateIntersection = do
+        points <- readLocalTip chainFollower
+        -- Cave: An empty list is interpreted as requesting the genesis point.
+        let points' = if null points
+                then [Point Origin]
+                else sortBy (flip compareSlot) points -- older points last
```

Comment:
> Ah wow — does this mean on master we'd always rollback to the oldest common checkpoint? 🤔

Fortunately not, as that would have meant rolling back to genesis (this is the issue that Piotr encountered). There was a call to `reverse` somewhere in the old `follow` function on master. 😅

Actually, the specification of the ChainSync mini-protocol stipulates that the node should ignore the sort order and returns the youngest point on the chain — but `cardano-node` behaves differently. See https://github.com/input-output-hk/ouroboros-network/issues/3443

> Is there a case for dropping the sortBy and changing the direction of the sort the DB returns (both wallet db and pool db)?

In light of the ChainSync spec, I would make the case for keeping the `sortBy` in the `chainSync` function and instead dropping the sort direction from the DB layer.

Reply:
> In light of the ChainSync spec, I would make the case for keeping the sortBy in the chainSync function and instead dropping the sort direction from the DB layer.

(But do nothing for now, as this would become obsolete in the DB layer redesign anyway.)